internal/pkg/agent/storage/store/migrations.go

```diff
@@ -0,0 +1,201 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package store
+
+import (
+	"errors"
+	"fmt"
+	"time"
+
+	"github.com/elastic/elastic-agent/internal/pkg/agent/storage"
+	"github.com/elastic/elastic-agent/internal/pkg/agent/storage/store/internal/migrations"
+	"github.com/elastic/elastic-agent/internal/pkg/fleetapi"
+	"github.com/elastic/elastic-agent/pkg/core/logger"
+)
+
+var ErrInvalidYAML = errors.New("could not parse YAML")
+
+func migrateActionStoreToStateStore(
+	log *logger.Logger,
+	actionStorePath string,
+	stateDiskStore storage.Storage) (err error) {
+
+	log = log.Named("state_migration")
+	actionDiskStore, err := storage.NewDiskStore(actionStorePath)
+	if err != nil {
+		return fmt.Errorf(
+			"could not create disk store when migratins action store: %w", err)
+	}
+
+	stateStoreExists, err := stateDiskStore.Exists()
+	if err != nil {
+		log.Errorf("failed to check if state store exists: %v", err)
+		return err
+	}
+
+	// do not migrate if the state store already exists
+	if stateStoreExists {
+		log.Debugf("state store already exists")
+		return nil
+	}
+
+	actionStoreExists, err := actionDiskStore.Exists()
+	if err != nil {
+		log.Errorf("failed to check if action store %s exists: %v", actionStorePath, err)
+		return err
+	}
+
+	// delete the actions store file upon successful migration
+	defer func() {
+		if err == nil && actionStoreExists {
```

Comment:
```suggestion
		if err == nil {
```

Can you get rid of this check if you put the defer after the `if !actionStoreExists` check below?